src/librustc_typeck/check/regionck.rs

```diff
@@ -759,6 +759,11 @@ impl<'a, 'gcx, 'tcx, 'v> Visitor<'v> for RegionCtxt<'a, 'gcx, 'tcx> {
             }
 
             hir::ExprRet(Some(ref ret_expr)) => {
+                if self.call_site_scope.is_none() {
+                    struct_span_err!(self.tcx.sess, expr.span, E0571,
+                                     "return statement cannot be out of a function scope").emit();
```

Comment:
How about "`return` can only be used inside a function body"?

It seems a bit odd to do this in `regionck` ... but I see that for (e.g.) a constant we use the type of the constant as the return type. That seems strange to me! I wonder how hard it would be to change that -- I'd rather have `FnCtxt` have a `Option<Ty<'tcx>>` as the return type, or at least do the flag in `librustc_typeck/check/mod.rs` and not `regionck`.

What do you think @eddyb ?